generate_workspace/src/main/java/com/google/devtools/build/workspace/maven/VersionResolver.java

```diff
@@ -0,0 +1,109 @@
+package com.google.devtools.build.workspace.maven;
+
+
+import com.google.common.collect.Lists;
+import org.eclipse.aether.RepositorySystem;
+import org.eclipse.aether.RepositorySystemSession;
+import org.eclipse.aether.artifact.Artifact;
+import org.eclipse.aether.repository.RemoteRepository;
+import org.eclipse.aether.resolution.VersionRangeRequest;
+import org.eclipse.aether.resolution.VersionRangeResolutionException;
+import org.eclipse.aether.resolution.VersionRangeResult;
+
+import java.util.List;
+
+import static com.google.devtools.build.workspace.maven.ArtifactBuilder.InvalidArtifactCoordinateException;
+
+/**
+ * Given a Maven coordinate with a version specification resolves the version of the coordinate in a
+ * similar fashion as Maven. Version specifications can include hard and soft pins as well as various
+ * forms of version ranges. When given a version range, Maven selects the highest available version.
+ * For both soft and hard pins, e.g. "4.2" or "[4.2]", Maven defaults to the pinned version.
+ */
+class VersionResolver {
+
+  private final List<RemoteRepository> remoteRepositories;
+  private final RepositorySystem repositorySystem;
+  private final RepositorySystemSession repositorySystemSession;
+
+  private VersionResolver(RepositorySystem repositorySystem,
+                          RepositorySystemSession systemSession,
+                          List<RemoteRepository> remoteRepositories) {
+    this.repositorySystem = repositorySystem;
+    this.remoteRepositories = remoteRepositories;
+    this.repositorySystemSession = systemSession;
+  }
+
+  /**
+   * Given a maven coordinate and its version specifications, selects the highest version if it is a version range
+   * or returns the pinned version if is a hard or soft pin.
+   */
+  String resolveVersion(String groupId, String artifactId, String versionSpec)
+      throws InvalidArtifactCoordinateException {
+
+    Artifact artifact = ArtifactBuilder.fromCoords(groupId, artifactId, versionSpec);
+    VersionRangeRequest rangeRequest = new VersionRangeRequest(artifact, remoteRepositories, null);
+
+    VersionRangeResult rangeResult;
+    try {
+      rangeResult = repositorySystem.resolveVersionRange(repositorySystemSession, rangeRequest);
+    } catch (VersionRangeResolutionException e) {
+      throw new InvalidArtifactCoordinateException(messageForInvalidArtifact(groupId, artifactId, versionSpec));
+    }
+
+    if (isInvalidRangeResult(rangeResult)) {
+      throw new InvalidArtifactCoordinateException(messageForInvalidArtifact(groupId, artifactId, versionSpec));
+    }
+    return rangeResult.getHighestVersion().toString();
```

Comment:
VersionRangeResult actually does not support either method. [[source](https://maven.apache.org/resolver/apidocs/org/eclipse/aether/resolution/VersionRangeResult.html)]

Those methods are associated with org.apache.maven.artifact.versioning.VersionRange. They are independent of aether. 

Let me play around with this.